design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
-    readinessChecks:
-    - type: None
+Starlark is a (very) limited dialect of Python designed for configuration. It
+embeds into many languages, including Go. This makes it a great candidate to
+build a generic Function that allows Composition authors to provide _inline_
+logic in a general-purpose-ish programming language. This is similar to the
+["GitHub Script"][github-script] GitHub Action that lets you plug some arbitrary
+logic into a GitHub Action pipeline when no existing Action does quite what you
+need.
+
+#### Iterating on Compositions
+
+Getting fast feedback on a Crossplane Composition has historically been a pain
+because the Composition logic has been encapsulated in a Kubernetes controller.
+In order to know whether your Composition works as intended you need to:
+
+1. Deploy (the right version of!) Crossplane to a Kubernetes cluster.
+2. Install Providers and Configurations for any resources you want to compose.
+3. Apply your XRD and Composition (potentially via a Configuration).
+4. Create an XR (or claim).
+5. See whether it works.
+
+Moving Composition logic out of Crossplane and into versioned Functions makes it
+a lot easier to test and iterate on, client-side. For example a `function test`
+command could test a single Function in isolation:
+
+```shell
+# Test a single Function by passing it a JSON RunFunctionRequest. This pulls and
+# starts the Function, makes the request, stops it, then returns the result.
+$ kubectl crossplane function test xpkg.upbound.io/negz/go-templates:v0.1.0 run-function-request.json
+{
+  # JSON encoded RunFunctionResponse omitted for brevity.
+}
 ```
 
-A function that wanted to set only an XR connection detail could return:
+You could similarly imagine a more holistic, `helm template`-like experience
+that locally rendered an entire Composition by pulling and running its templates
+locally:
 
-```yaml
-apiVersion: apiextensions.crossplane.io/v1alpha1
-kind: FunctionIO
-observed: {}  # Omitted for brevity.
-desired:
-  composite:
-    connectionDetails:
-    - type: FromValue
-      name: username
-      value: admin
+```shell
+$ kubectl crossplane composition render composition.yaml xr.yaml
+---
+# YAML stream omitted for brevity.
 ```
 
-A function wishing to delete a composed resource may do so by setting its
-`resource` to null, for example:
+Assuming a 'pure' Composition consisting only of Functions with no P&T
+resources, this command would need only to iterate through the Composition's
+`functions` array, and for each Function:
+
+1. Pull and start it.
+2. Form a `RunFunctionRequest` from the XR and any prior `RunFunctionResponse`.
+3. Print the `desired` block of final `RunFunctionResponse` as a YAML stream.
+
+This isn't quite an end-to-end test of Composition. The composed resources are
+not actually created and reconciled with an external system, so its not possible
+for example to derive XR status from composed resource status. It's also not
```

Comment:
> The composed resources are not actually created and reconciled with an external system, so its not possible for example to derive XR status from composed resource status

Wondering if this could be mocked by providing an `observed.yaml` as yet another input to the render command.

Reply:
Hrm, potentially. It would be neat to do at least optionally. I'll update the doc to mention it.